File: DictionaryProject/Commands/TranslateCommand.vb
Imports System.Text.RegularExpressions

<Highlight(Regex:="^\s*translate \\\{", Color:="Green", NextLevel:=2)> _
<Highlight(Regex:="#|=>", Color:="Red", Level:=2)> _
<Highlight(Regex:="\\\}", Color:="Green", Level:=2, NextLevel:=-2)> _
Public Class TranslateCommand
    Inherits CommandBase

    Private _stringmapping As New Dictionary(Of String(), String())
    Private _mapping As New Dictionary(Of Regex, String())

    Public Overrides Sub CheckSanity(ByVal project As Project)
        If _stringmapping.Count = 0 Then
            project.Warnings.Add(String.Format("Line {0}: The Translate command requires at least one translation directive.", LineNumber))
        End If

        For Each key As String() In _stringmapping.Keys
            Dim reg As Regex = MakeRegex(key)
            _mapping.Add(reg, _stringmapping(key))
        Next
    End Sub

    Private Function ExpandTokenSet(ByVal token As String, ByVal ingroup As Boolean) As String
        If token.StartsWith("$") Then
            Dim sb As New System.Text.StringBuilder()

            If Not ingroup Then
                sb.Append("(?:")
            End If

            For Each tk As String In Project.TokenSets.FindByName(token.Substring(1)).Tokens
                sb.Append(tk)
                sb.Append("|")
            Next

            If Not ingroup Then
                sb.Append(")")
            End If

            Return sb.ToString()
        End If

        Return token
    End Function

    Private Function MakeRegex(ByVal source As String()) As Regex
        Dim sb As New System.Text.StringBuilder()

        If source.First = "#" Then
            sb.Append("^")
            source = source.Skip(1).ToArray()
        End If

        sb.Append("\|")

        Dim ingroup As Boolean = False

        For Each token As String In source
            Dim bPickup As Boolean = False
            If token.StartsWith("(") AndAlso token.EndsWith(")") Then
                bPickup = True
                token = token.Substring(1, token.Length - 2)
                sb.Append("(")
            End If

            If token.StartsWith("[") Then
                sb.Append("(?:")

                For Each tk As String In ProjectSerializer.ReadTokens(token.Substring(1, token.Length - 2))
                    sb.Append(ExpandTokenSet(tk, True))
                    sb.Append("|")
                Next

                sb.Remove(sb.Length - 1, 1)
                sb.Append(")")
                ingroup = False
            ElseIf token = "#" Then
                'sb.Remove(sb.Length - 2, 2)
                sb.Append("$")
                Return New Regex(sb.ToString())
            Else
                sb.Append(ExpandTokenSet(token, False))
            End If

            If bPickup Then
                sb.Append(")")
            End If

            sb.Append("\|")
        Next

        'If sb.Length > 2 Then
        '    sb.Remove(sb.Length - 2, 2)
        'End If

        Debug.WriteLine(sb.ToString())
        Return New Regex(sb.ToString())
    End Function

    Private Function ExpandReplacement(ByVal destination As String(), ByVal m As Match) As String
        Dim result As New List(Of String)()

        For Each token As String In destination
            Dim tokenId As Integer
            If token.StartsWith("\") AndAlso Integer.TryParse(token.Substring(1), tokenId) Then
                result.Add(m.Groups(tokenId).Value)
            ElseIf token.StartsWith("$") Then
                result.Add(Project.TokenSets.FindByName(token.Substring(1)).GetToken())
            ElseIf token.StartsWith("[") Then
                Dim tokens As List(Of String) = ProjectSerializer.ReadTokens(token.Substring(1, token.Length - 2))

                Dim ts As New TokenSet()

                For Each t As String In tokens
                    ts.Tokens.AddRange(ExpandTokenSet(t, True).Split("|"c))
                Next

                result.Add(ts.GetToken())
            Else
                result.Add(token)
            End If
        Next

        Return "|" & String.Join("|", result.ToArray()) & "|"
    End Function

    Public Overrides Sub Execute(ByVal context As Context)
        Dim input As String = "|" & String.Join("|", context.Tokens.Where(Function(token As String) Not String.IsNullOrEmpty(token)).ToArray()) & "|"

        For Each regex As Regex In _mapping.Keys
            Dim destination As String() = _mapping(regex)

            input = regex.Replace(input, Function(m As Match) ExpandReplacement(destination, m))
        Next

        context.Tokens.Clear()
        context.Tokens.AddRange(input.Split(New Char() {"|"c}, StringSplitOptions.RemoveEmptyEntries))
    End Sub

    Public Overrides Sub LoadCommand(ByVal project As Project, ByVal reader As System.IO.TextReader, ByVal line As String, ByRef lineNumber As Integer)
        MyBase.LoadCommand(project, reader, line, lineNumber)

        Dim parts As List(Of String) = ProjectSerializer.ReadTokens(line)

        If parts.Count = 2 AndAlso parts(1) = "{" Then
            If ProjectSerializer.ReadLines(project, reader, AddressOf ReadTranslation, lineNumber) Then
                project.Warnings.Add(String.Format("Line {0}: The Translate command is not closed correctly.", lineNumber))
            End If
        Else
            project.Warnings.Add(String.Format("Line {0}: The Translate command requires a {{} at the end of the line.", lineNumber))
        End If
    End Sub

    Private Function ReadTranslation(ByVal context As Object, ByVal reader As IO.TextReader, ByVal line As String, ByRef lineNumber As Integer) As Boolean
        If line = "}" Then
            Return False
        End If

        If String.IsNullOrEmpty(line) Then
            Return True
        End If

        Dim tokens As List(Of String) = ProjectSerializer.ReadTokens(line)

        Dim source As New List(Of String)()
        Dim dest As New List(Of String)()

        Dim translatepos As Integer = tokens.IndexOf("=>")

        If translatepos = -1 Then
            Project.Warnings.Add(String.Format("Line {0}: Missing => in Translate directive.", lineNumber))
        ElseIf translatepos = 0 Then
            Project.Warnings.Add(String.Format("Line {0}: Missing source tokens in Translate directive.", lineNumber))
        Else
            _stringmapping.Add(tokens.Take(translatepos).ToArray(), tokens.Skip(translatepos + 1).ToArray())
        End If

        Return True
    End Function

    Public Overrides Sub WriteCommand(ByVal writer As System.IO.TextWriter)
        writer.WriteLine("Translate {")
        For Each key As String() In _stringmapping.Keys
            writer.WriteLine("{0} => {1}", ProjectSerializer.SecureList(key), ProjectSerializer.SecureList(_stringmapping(key)))
        Next
        writer.WriteLine("}")
    End Sub
End Class

File: monowordbuilder/wordbuilderbase/Commands/TranslateCommand.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace Whee.WordBuilder.Model.Commands
{
    [Highlight(RegEx = "^\\s*translate \\\\\\{", Color = "Green", NextLevel = 2)]
    [Highlight(RegEx = "#|=>", Color = "Red", Level = 2)]
    [Highlight(RegEx = "\\\\\\}", Color = "Green", Level = 2, NextLevel = -2)]
    public class TranslateCommand : CommandBase
    {

        private Dictionary<string[], string[]> _stringmapping = new Dictionary<string[], string[]>();
        private Dictionary<Regex, string[]> _mapping = new Dictionary<Regex, string[]>();

        private string ExpandTokenSet(string token, bool ingroup)
        {
            if (token.StartsWith("$"))
            {
                System.Text.StringBuilder sb = new System.Text.StringBuilder();

                if (!ingroup)
                {
                    sb.Append("(?:");
                }

                foreach (string tk in Project.TokenSets.FindByName(token.Substring(1)).Tokens)
                {
                    sb.Append(tk);
                    sb.Append("|");
                }

                if (!ingroup)
                {
                    sb.Append(")");
                }

                return sb.ToString();
            }

            return token;
        }

        private Regex MakeRegex(string[] sourceOrig)
        {
            string[] source;
            System.Text.StringBuilder sb = new System.Text.StringBuilder();

            if (sourceOrig[0] == "#")
            {
                sb.Append("^");
                source = new string[sourceOrig.Length - 2];
                Array.Copy(sourceOrig, 1, source, 0, sourceOrig.Length - 1);
            }
            else
            {
                source = sourceOrig;
            }

            sb.Append("\\|");

            //		bool ingroup = false;

            foreach (string tokeniter in source)
            {
                bool bPickup = false;
                string token = tokeniter;
                if (token.StartsWith("(") && token.EndsWith(")"))
                {
                    bPickup = true;
                    token = token.Substring(1, token.Length - 2);
                    sb.Append("(");
                }

                if (token.StartsWith("["))
                {
                    sb.Append("(?:");

                    foreach (string tk in ProjectSerializer.ReadTokens(token.Substring(1, token.Length - 2)))
                    {
                        sb.Append(ExpandTokenSet(tk, true));
                        sb.Append("|");
                    }

                    sb.Remove(sb.Length - 1, 1);
                    sb.Append(")");
                }
                else if (token == "#")
                {
                    //sb.Remove(sb.Length - 2, 2)
                    sb.Append("$");
                    return new Regex(sb.ToString());
                }
                else
                {
                    sb.Append(ExpandTokenSet(token, false));
                }

                if (bPickup)
                {
                    sb.Append(")");
                }

                sb.Append("\\|");
            }

            //If sb.Length > 2 Then
            //    sb.Remove(sb.Length - 2, 2)
            //End If

            Debug.WriteLine(sb.ToString());
            return new Regex(sb.ToString());
        }

        private string ExpandReplacement(string[] destination, Match m)
        {
            List<string> result = new List<string>();

            foreach (string token in destination)
            {
                int tokenId;
                if (token.StartsWith("\\") && int.TryParse(token.Substring(1), out tokenId))
                {
                    result.Add(m.Groups[tokenId].Value);
                }
                else if (token.StartsWith("$"))
                {
                    result.Add(Project.TokenSets.FindByName(token.Substring(1)).GetToken());
                }
                else if (token.StartsWith("["))
                {
                    List<string> tokens = ProjectSerializer.ReadTokens(token.Substring(1, token.Length - 2));

                    TokenSet ts = Project.CreateTokenSet();

                    foreach (string t in tokens)
                    {
                        ts.Tokens.AddRange(ExpandTokenSet(t, true).Split('|'));
                    }

                    result.Add(ts.GetToken());
                }
                else
                {
                    result.Add(token);
                }
            }

            return "|" + string.Join("|", result.ToArray()) + "|";
        }

        public override void Execute(Context context)
        {
            string input = "|" + string.Join("|", context.Tokens.FindAll((string token) => !string.IsNullOrEmpty(token)).ToArray()) + "|";

            foreach (Regex regex in _mapping.Keys)
            {
                string[] destination = _mapping[regex];

                input = regex.Replace(input, (Match m) => ExpandReplacement(destination, m));
            }

            context.Tokens.Clear();
            context.Tokens.AddRange(input.Split(new char[] { '|' }, StringSplitOptions.RemoveEmptyEntries));
        }

        public override void LoadCommand(Whee.WordBuilder.ProjectV2.IProjectSerializer serializer)
        {
            m_lineNumber = serializer.LineNumber;

            ProjectV2.TranslationBlockNode tbn = new ProjectV2.TranslationBlockNode(serializer);

            Children.Add(tbn);

            foreach (ProjectV2.TranslationNode item in tbn.Children)
            {
                _stringmapping.Add(item.Source.ToArray(), item.Destination.ToArray());
            }
        }

        public override void LoadCommand(Project project, System.IO.TextReader reader, string line, ref int lineNumber)
        {
            base.LoadCommand(project, reader, line, ref lineNumber);

            List<string> parts = ProjectSerializer.ReadTokens(line);

            if (parts.Count == 2 && parts[1] == "{")
            {
                if (ProjectSerializer.ReadLines(project, reader, ReadTranslation, ref lineNumber))
                {
                    project.Warnings.Add(string.Format("Line {0}: The Translate command is not closed correctly.", lineNumber));
                }
            }
            else
            {
                project.Warnings.Add(string.Format("Line {0}: The Translate command requires a {{} at the end of the line.", lineNumber));
            }
        }

        private bool ReadTranslation(object context, TextReader reader, string line, ref int lineNumber)
        {
            if (line == "}")
            {
                return false;
            }

            if (string.IsNullOrEmpty(line))
            {
                return true;
            }

            List<string> tokens = ProjectSerializer.ReadTokens(line);

            int translatepos = tokens.IndexOf("=>");

            if (translatepos == -1)
            {
                Project.Warnings.Add(string.Format("Line {0}: Missing => in Translate directive.", lineNumber));
            }
            else if (translatepos == 0)
            {
                Project.Warnings.Add(string.Format("Line {0}: Missing source tokens in Translate directive.", lineNumber));
            }
            else
            {
                _stringmapping.Add(tokens.GetRange(0, translatepos).ToArray(), tokens.GetRange(translatepos + 1, tokens.Count - translatepos - 1).ToArray());
            }

            return true;
        }

        public override void WriteCommand(System.IO.TextWriter writer)
        {
            writer.WriteLine("Translate {");
            foreach (string[] key in _stringmapping.Keys)
            {
                writer.WriteLine("{0} => {1}", ProjectSerializer.SecureList(key), ProjectSerializer.SecureList(_stringmapping[key]));
            }
            writer.WriteLine("}");
        }

        public override void CheckSanity(Project project, Whee.WordBuilder.ProjectV2.IProjectSerializer serializer)
        {
            m_project = project;
            if (_stringmapping.Count == 0)
            {
                serializer.Warn("The Translate command requires at least one translation directive.", this);
            }

            foreach (string[] key in _stringmapping.Keys)
            {
                Regex reg = MakeRegex(key);
                _mapping.Add(reg, _stringmapping[key]);
            }
        }

        public override bool RequireNewLine
        {
            get
            {
                return false;
            }
        }
    }
}